### PR TITLE
Add new sygus grammar construction utility

### PR DESCRIPTION
**Pull request by ajreynol**
This utility will replace https://github.com/cvc5/cvc5/blob/main/src/theory/quantifiers/sygus/sygus_grammar_cons.h.

The main feature is that the new construction is based on `SygusGrammar` and allows the user to get the unresolved sygus grammar for fine-grained modifications.

Note that much of the code is heavily based on the old utility, but done in a cleaner way that uses sygus grammar instead of sygus datatypes.

Review https://github.com/cvc5/cvc5/pull/9817 first.

Further PRs will change callers to use the new utility.

### CODE REVIEW

File: src/theory/quantifiers/sygus/sygus_grammar_cons_new.h [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+   * grammar.
+   * @param bvl The input variable list for the grammar (if it exists). If
+   * provided, each variable in this list is added as a non-terminal rule.
+   */
+  static TypeNode mkDefaultSygusType(const Options& opts,
+                                     const TypeNode& range,
+                                     const Node& bvl);
+  /**
+   * Make the type corresponding to the default sygus grammar type returning
+   * range with bound variables bvl, with a custom set of initial non-terminal
```
Comment (abdoo8080) [inline src/theory/quantifiers/sygus/sygus_grammar_cons_new.h:54]: Why non-terminal rules? Shouldn't they be terminal rules?

File: src/theory/quantifiers/sygus/sygus_grammar_cons_new.h [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+   * @param opts The options, which impact grammar construction
+   * @param range The type of terms generated by the start symbol of the
+   * grammar.
+   * @param bvl The input variable list for the grammar (if it exists).
+   * @param trules The provided set of non-terminal rules.
+   */
+  static SygusGrammar mkDefaultGrammar(const Options& opts,
+                                       const TypeNode& range,
+                                       const Node& bvl,
+                                       const std::vector<Node>& trules);
```
Comment (abdoo8080) [inline src/theory/quantifiers/sygus/sygus_grammar_cons_new.h:95]: ditto

File: src/theory/quantifiers/sygus/sygus_grammar_cons_new.h [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+   * @param opts The options, which impact grammar construction
+   * @param range The type of terms generated by the start symbol of the
+   * grammar.
+   * @param bvl The input variable list for the grammar (if it exists).
+   * @param trules The provided set of non-terminal rules.
+   */
+  static SygusGrammar mkEmptyGrammar(const Options& opts,
+                                     const TypeNode& range,
+                                     const Node& bvl,
+                                     const std::vector<Node>& trules);
```
Comment (abdoo8080) [inline src/theory/quantifiers/sygus/sygus_grammar_cons_new.h:121]: ditto

**Review by abdoo8080 — CHANGES_REQUESTED** [review]

File: src/theory/quantifiers/sygus/sygus_grammar_cons_new.cpp [hunk 4]
```diff
@@ -0,0 +1,949 @@
...
+  // get the variables
+  std::vector<Node> vars;
+  if (!bvl.isNull())
+  {
+    Assert(bvl.getKind() == BOUND_VAR_LIST);
+    vars.insert(vars.end(), bvl.begin(), bvl.end());
+  }
+  // collect the types we are considering, which is all component types of
+  // the range type and the initial terminal rules. We also always include
+  // Bool.
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/sygus/sygus_grammar_cons_new.cpp:134]: Bound variables are typically terminal rules.

File: src/theory/quantifiers/sygus/sygus_grammar_cons_new.h [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+  /**
+   * Make the type corresponding to the default sygus grammar type returning
+   * range with bound variables bvl, with a custom set of initial non-terminal
+   * rules.
+   *
+   * @param opts The options, which impact grammar construction
+   * @param range The type of terms generated by the start symbol of the
+   * grammar.
+   * @param bvl The input variable list for the grammar (if it exists).
+   * @param trules The provided set of non-terminal rules.
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/quantifiers/sygus/sygus_grammar_cons_new.h:61]: It should say "provided set of terminal rules".

File: src/theory/quantifiers/sygus/sygus_grammar_cons_new.h (hunk 1 above)
Comment (ajreynol) [reply src/theory/quantifiers/sygus/sygus_grammar_cons_new.h:54]: Yes

**Review by abdoo8080 — APPROVED** [review]

**Pull request merged**